File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    pair<int,int> recur(TreeNode* node,bool& ans) {
        if(!node)
            return {INT_MAX,INT_MIN};
        pair<int,int> l=recur(node->left,ans),
                    r=recur(node->right,ans);
        if(l.second>=(node->val) || r.first<=(node->val))
            ans=false;
        int small=min(node->val,min(l.first,r.first)),
            large=max(node->val,max(l.second,r.second));
        return {small,large};
    }
    bool isValidBST(TreeNode* root) {
        bool ans=true;
        recur(root,ans);
        return ans;
    }
};

File: Data Structures & Algorithms/longest-consecutive-sequence/submission-4.js
class Solution {
    max(a,b) {
        if(a>=b) return a;
        return b;
    }
    longestConsecutive(nums) {
        if(nums.length==0) return 0;
        nums.sort((a, b) => {
            if(a<b) return -1;
            else if(b<a) return 1;
            return 0;
        });
        // console.log(nums);
        let ans=1;
        const mp = new Map();
        for(let i=0;i<nums.length;i++) {
            let val=1;
            if(mp.has(nums[i]-1)) {
                val+=mp.get(nums[i]-1);
            }
            if(mp.has(nums[i]))
                mp.set(nums[i],this.max(mp.get(nums[i]),val));
            else
                mp.set(nums[i],val);
            if(val>ans) ans=val;
        }
        return ans;
    }
}